File: java/Dado.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.Random;
import civitas.Diario;


/**
 *
 * @author carlos
 */
public class Dado {
    
    static final private Dado instance = new Dado();
    
    private Random random;
    
    private int ultimoResultado;
    private boolean debug;
    
    private static int SalidaCarcel = 5;
    private static int numCaras = 6;
    
    
    //Constructor
    private Dado(){
        debug = false;
        random = new Random();       
        
    }
    
    //Métodos
    
    int tirar(){
        if (!debug){
            ultimoResultado = random.nextInt(numCaras) + 1;
        }
        else{
            ultimoResultado = 1;
        }
        
        return ultimoResultado;
    }
    
    
    boolean salgoDeLaCarcel(){
        boolean salir = false;
        
        if (ultimoResultado >= 5){
            salir = true;
        }
        
        return salir;
    }
    
    
    int quienEmpieza(int n){
        int jugador = random.nextInt(n);
        
        return jugador;
    }
    
    void setDebug(boolean d){
        //Añadir registro en el diario
        debug = d;
        
        String evento;
        
        if (d){
            evento = "Dado modo Debug";
        }
        else{
            evento = "Dado modo normal";
        }
        
        Diario.getInstance().ocurreEvento(evento);
    }
    
    int getUltimoResultado(){
        return ultimoResultado;
    }
    
    public static Dado getInstance(){
		return instance;
	}
    
}

File: ruby/dado.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module Civitas
  class Dado
    attr_reader :random , :ultimo_resultado
    attr_writer :debug
    
    
    @num_caras=6
    
    def initialize
       @debug = false
       @ultimo_resultado=0
       @random = Random.new
    end
    
    @@instance = Dado.new
    
    def tirar()
      resultado = 1
      if(!@debug)
        resultado = @random.rand(6) + 1
      end
      @ultimo_resultado =resultado
      return @ultimo_resultado
    end
    
    def salir_de_la_carcel()
      
      if(tirar() >= 5)
        return true
      else
        return false
      end
  
    end
    
    def quien_empieza(n)
      return @random.rand(n)
    end
  
    def set_debug(d)
      String modo = "desactivado";
      if(d == true)
        modo = "Activado"
      end
      Diario.instance.ocurre_evento("Modo debug: #{modo}")
      @debug = d
    end
    
    def get_ultimo_resultado()
      return @ultimo_resultado
    end
  end
end